fix(nav-bar): improve screen reader support

MarcoZehe opened the pull request:
… reader users

* make the nav-bar component a tablist.
* make the nav-item's li element presentational and the button a tab.
* apply the aria-label, if provided, to the button directly, not the li element.

Fixes #11485

<!-- 
Filling out this template is required! Do not delete it when submitting a Pull Request! Without this information, your Pull Request may be auto-closed.
-->
## PR Checklist
Please check that your PR fulfills the following requirements:
- [x] The commit message follows [our guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format)
- [x] Tests for the changes have been added or this is not a bug fix / enhancement
- [x] Docs have been added, updated, or were not required

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Enhancement
[ ] Documentation content changes
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying and link to one or more relevant issues. -->
Issue Number: #11485 

Currently, the nav-bar element is exposed as a listbox accessible to screen reader users. Its nav-item child consists of a li element which currently has role "option" (as appropriate for a listbox child), and the button inside that li is exposed as is.

Screen readers do have a hard time interacting with this construct, since they do not expect an option to contain a button.

## What is the new behavior?

The new behavior is that the nav-bar element is exposed as a tablist, and the button inside the li of the nav-item as a tab accessible. In addition, the aria-label that was originally applied to the nav-item's li element is now applied to the buton, since the li in the tablist/tab pattern is to be marked presentational. aria-label must therefore go onto the element that has role "tab".

No changes besides semantic ARIA exposure has been changed, the keyboard behavior is the same as well as any other behavior this component exposes. Web developers needn't change anything, either.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
<!-- Note that breaking changes are highly unlikely to get merged to master unless the validation is clear and the use case is critical. -->

## Other information

The tests have been adjusted to look for the aria-label on the tab directly, not the parent.

Splaktar commented:
The build failure looks like a fluke. I have restarted it.

Splaktar commented:
It looks like the test failures are not a fluke as they keep happening and I can reproduce them on my local machine. Are you able to get all of the tests to pass on your local machine with `gulp karma`?

Splaktar commented:
I've opened https://github.com/angular/material/issues/11487 to track the test failures on `master`.

MarcoZehe commented on src/components/navBar/navBar.js:
```diff
@@ -457,6 +459,7 @@ function MdNavItem($mdAria, $$rAF, $mdUtil, $window) {
       'mdNavSref': '@?',
       'srefOpts': '=?',
       'name': '@',
+      'navItemAriaLabel': '@?',
```
Hm, what do you think needs updating? It only says that this adds alternative text for accessibility, and that is still a valid statement. It only internally adds it in a different spot.

MarcoZehe replied:
OK, I've made it explicit that the aria-label is applied to the `<md-button>` element.

Splaktar commented on src/components/navBar/navBar.js:
```diff
@@ -90,7 +90,7 @@ angular.module('material.components.navBar', ['material.core'])
  * to be specified.
  *
- * @param {string=} aria-label Adds alternative text for accessibility.
+ * @param {string=} aria-label Adds alternative text for accessibility to the `<md-button>` element.
```
I think that my main point was that we need to rename the API from `aria-label` to `nav-item-aria-label`.

For the text, let's use "Allows setting or overriding the label that is announced by a screen reader for the nav item's button. If this is not set, the nav item's transcluded content will be announced. Make sure to set this if the nav item's transcluded content does not include descriptive text, for example only an icon."

MarcoZehe commented:
@Splaktar Thanks for the review! I also changed the documentation further up to point to the tablist and tab specifications instead of the outdated site navigator pattern.

Splaktar commented on src/components/navBar/navBar.js:
```diff
@@ -498,7 +504,9 @@ function MdNavItem($mdAria, $$rAF, $mdUtil, $window) {
         }
 
-        $mdAria.expectWithText(element, 'aria-label');
+        if (!ctrl.navItemAriaLabel) {
```
`ctrl` is not defined at this point. `mdNavItem` should be used instead.

Splaktar commented on src/components/navBar/navBar.js:
```diff
@@ -127,7 +130,7 @@ function MdNavBar($mdAria, $mdTheming) {
       '<div class="md-nav-bar">' +
         '<nav role="navigation">' +
-          '<ul class="_md-nav-bar-list" ng-transclude role="listbox" ' +
+          '<ul class="_md-nav-bar-list" ng-transclude role="tablist" ' +
```
The docs on line 46 still refer to `listbox` and should be updated to tablist.

Splaktar replied:
Similarly, as you noted in a comment, the docs on lines 34-41, need to be updated to point to the applicable tablist specs.

Splaktar commented on src/components/navBar/navBar.js:
```diff
@@ -438,6 +441,9 @@ function MdNavItem($mdAria, $$rAF, $mdUtil, $window) {
             'ng-disabled="ctrl.disabled" ' +
             'tabindex="-1" ' +
+            'role="tab" ' +
+            'aria-label="{{ctrl.navItemAriaLabel}}" ' +
```
We don't want to set an empty `aria-label` here as that will be flagged as an invalid aria attribute and may cause issues with certain screen readers.

If no `nav-item-aria-label` is defined, no `aria-label` attribute should be applied. We can do this using the following:
```
'ng-attr-aria-label="{{ctrl.navItemAriaLabel ? ctrl.navItemAriaLabel : undefined}}" ' +
```
I have a few examples of this in https://github.com/angular/material/pull/11441/files#diff-dfc7d490c731a45e66ad22d8efb7456dR358.

Splaktar commented on src/components/navBar/navBar.js:
```diff
@@ -498,7 +504,9 @@ function MdNavItem($mdAria, $$rAF, $mdUtil, $window) {
 
-        $mdAria.expectWithText(element, 'aria-label');
+        if (!ctrl.navItemAriaLabel) {
+          $mdAria.expectWithText(element, 'aria-label');
```
This is still checking for the `aria-label` on the `li` element. Replacing `element` with `navButton` here seems to properly check the `md-button._md-nav-button` element.

Splaktar commented on src/components/navBar/navBar.spec.js:
```diff
@@ -364,13 +364,13 @@ describe('mdNavBar', function() {
     it('does not change aria-label on nav items', function() {
       createTabs();
-      expect(getTab('tab3').parent().attr('aria-label')).toBe('foo');
+      expect(getTab('tab3').attr('aria-label')).toBe('foo');
```
To ensure that we don't break any apps, let's continue to check that the `aria-label` API still works the way it did before:
```js
    it('does not change aria-label on nav items', function() {
      createTabs();
      expect(getTab('tab3').parent().attr('aria-label')).toBe('foo');
    });
```
Then let's add a new check for the `nav-item-aria-label`
```js
    it('does not change nav-item-aria-label on nav item buttons', function() {
      createTabs();
      expect(getTab('tab4').attr('aria-label')).toBe('foo');
    });
```
This will also require adding another tab to `createTabs()` around line 47:
```js
        '  <md-nav-item md-nav-href="#4" name="tab4" nav-item-aria-label="foo">' +
        '    tab4' +
        '  </md-nav-item>' +
```

Splaktar commented:
I've tracked down the CI failure and the disconnection issue to [this test](https://github.com/angular/material/blob/master/src/components/navBar/navBar.spec.js#L124-L132).

MarcoZehe commented:
> I've opened PR #11488 to fix the test issues here. Sorry that you had to run into them!

Thank you for this, and for your patience and guidance with my PR! I think I've addressed all your comments and will push an updated revision in a minute.

Splaktar commented:
@MarcoZehe OK, PR #11488 has been merged. Can you please rebase this and then force push to your branch? That should hopefully get all of the tests passing in TravisCI. Thank you!

MarcoZehe commented:
> 
> 
> @MarcoZehe OK, PR #11488 has been merged. Can you please rebase this and then force push to your branch? That should hopefully get all of the tests passing in TravisCI. Thank you!

Yes, all passed! \-O-/

Splaktar commented:
The first line of your commit is a bit too long. Can you please update it to just be "fix(nav-bar): improve screen reader support"? Thanks!

MarcoZehe commented:
> 
> 
> The first line of your commit is a bit too long. Can you please update it to just be "fix(nav-bar): improve screen reader support"? Thanks!

Done. :) Thank you again for your patience and guidance!

Splaktar commented:
Looks good, thank you! I'll let you know if we run into any issues in the "presubmit" process. This will run the change through all of the relevant tests in Google to make sure that it passes unit tests, screenshot tests, and a11y tests.

Splaktar commented:
I've submitted PR https://github.com/angular/material/pull/11494 which should improve the a11y state of nav-bar quite a bit more. I tested it with the changes in this PR and it worked well and doesn't seem to conflict.
